internal/federation/server_room.go

```diff
@@ -97,24 +97,38 @@ func (r *ServerRoom) AuthChain() (chain []*gomatrixserverlib.Event) {
 	return r.AuthChainForEvents(r.AllCurrentState())
 }
 
-// AuthChainForEvents returns all auth events for all events in the given state TODO: recursively
+// AuthChainForEvents returns all auth events for all events in the given state
 func (r *ServerRoom) AuthChainForEvents(events []*gomatrixserverlib.Event) (chain []*gomatrixserverlib.Event) {
 	chainMap := make(map[string]bool)
-	// get all the auth event IDs
-	for _, ev := range events {
+
+	// build a map of all events in the room
+	// Timeline and State contain different sets of events, so check them both.
+	eventsByID := map[string]*gomatrixserverlib.Event{}
+	for _, ev := range r.Timeline {
+		eventsByID[ev.EventID()] = ev
+	}
+	for _, ev := range r.State {
+		eventsByID[ev.EventID()] = ev
+	}
+
+	// a queue of events whose auth events are to be included in the auth chain
+	queue := []*gomatrixserverlib.Event{}
+	queue = append(queue, events...)
+
+	// get all the auth events recursively
+	// we extend the "queue" as we go along
+	for i := 0; i < len(queue); i++ {
+		ev := queue[i]
 		for _, evID := range ev.AuthEventIDs() {
 			if chainMap[evID] {
 				continue
 			}
 			chainMap[evID] = true
+			chain = append(chain, eventsByID[evID])
+			queue = append(queue, eventsByID[evID])
 		}
 	}
-	// find them in the timeline
-	for _, tev := range r.Timeline {
-		if chainMap[tev.EventID()] {
-			chain = append(chain, tev)
-		}
-	}
+
 	return
 }
```

Comment:
Without this change, the complement homeserver excludes hs1's membership event from the auth chain sent to hs2. hs2 is unhappy and tries to make `/event_auth` requests, which the complement homeserver is not set up to handle.